### ISSUE TITLE
Add the ability the write messages with a partition key

### ISSUE THREAD

**Issue by hammond756** [Nordstrom/kafka-connect-sqs#38]
Our Kafka topics are set up with a key-based retention policy. The keys we want to use correspond to values from the payload on the SQS messages. Currently we have set up intermediate topics that the connector writes to and consumer/producers that read the messages and write to another topic with the correct partition key.

Is there appetite for a configuration of the connect agent that can set the partition key? In our case the message payload is JSON and I know that's not always the case. But I can see an implementation that relies on the MessageAttributes of the SQS messages to be useful and more generally applicable.

It's been a long time since I've written any Java but I'd be willing to try my hand at a PR if the feature is something could be accepted.

Cheers!

**Comment by dylanmei** [comment]
I'll admit I don't understand the use case and feature you are describing. Does "key-based retention policy" mean the topics are compacted? Do you wish to produce data into Kafka using a MessageAttribute as message key?

**Comment by hammond756** [comment]
Yes, that’s correct. The topics are compacted and I want to produce to Kafka using a MessageAttribute as key. 

I also shared this issue with a colleague and he mentioned that such a thing should already be possible by using a Single Message Transformation. I tried setting up the `valueToKey` transform but now I’m running into the problem that I’ve set up the connectors using a string encoder. 

I’ll definitely be spending a bit more time on the SBT route as that would mean we rely on established functionality instead of building new ones.

**Comment by dylanmei** [comment]
Currently, the Kafka message key is simply the SQS message ID.

VaultToKey transform may be useful if the value you want to key by is within the message body.

If it only exists as a MessageAttribute, some treatment would be required within in [SqsSourceConnectorTask.java](https://github.com/Nordstrom/kafka-connect-sqs/blob/master/src/main/java/com/nordstrom/kafka/connect/sqs/SqsSourceConnectorTask.java#L99):
- an extraction of the attribute you want to key by
- handling when that attribute is not present in the message

**Comment by GDegrove** [comment]
Hello @dylanmei , 

Thanks for the reply. We were looking at this issue, and unfortunately we cannot use the SMT as the `ConnectSourceRecord` that is being produced by the connector is always of type `String`. Rather than a Struct. 

To be able to use the Connector with the SMT's it would be great to be able to convert the record to type Struct (if configured so)

Is there a plan to support Struct as `SourceRecord`? That would be awesome as this will enable the connector to be used more freely in the actual kafka-connect ecosystem (SMT + converter)

**Comment by hammond756** [comment]
@GDegrove a related issue was opened for the SqsSinkConnectorTask: https://github.com/Nordstrom/kafka-connect-sqs/issues/39

**Comment by dylanmei** [comment]
There hasn't been a plan in our org but it is an obvious shortcoming to not property support any converters.

**Comment by koosie0507** [comment]
@dylanmei I've created two pull requests (linked) that should solve this issue one way or the other. Hope we can merge one of them. Would you be up to review them?

**Comment by dylanmei** [comment]
@koosie0507 I appreciate how #44 safely avoids digging through the message contents by utilizing features of SQS. 👍

**Comment by koosie0507** [comment]
Great! I followed up on your comments on that PR and closed #43.
